manifest/oci_index.go

```diff
@@ -69,36 +76,60 @@ func (index *OCI1Index) UpdateInstances(updates []ListUpdate) error {
 	return nil
 }
 
-// ChooseInstance parses blob as an oci v1 manifest index, and returns the digest
-// of the image which is appropriate for the current environment.
-func (index *OCI1Index) ChooseInstance(ctx *types.SystemContext) (digest.Digest, error) {
-	wantedPlatforms, err := platform.WantedPlatforms(ctx)
-	if err != nil {
-		return "", fmt.Errorf("getting platform information %#v: %w", ctx, err)
-	}
-	for _, wantedPlatform := range wantedPlatforms {
-		for _, d := range index.Manifests {
+func instanceBetterMatchesPlatforms(ctx *types.SystemContext, manifests []imgspecv1.Descriptor, platformCheck bool, wantedPlatform imgspecv1.Platform) digest.Digest {
+	var matchedDigest digest.Digest
+	matchedDigest = ""
+	for _, d := range manifests {
+		var imagePlatform imgspecv1.Platform
+		if platformCheck {
 			if d.Platform == nil {
 				continue
 			}
-			imagePlatform := imgspecv1.Platform{
+			imagePlatform = imgspecv1.Platform{
 				Architecture: d.Platform.Architecture,
 				OS:           d.Platform.OS,
 				OSVersion:    d.Platform.OSVersion,
 				OSFeatures:   dupStringSlice(d.Platform.OSFeatures),
 				Variant:      d.Platform.Variant,
 			}
-			if platform.MatchesPlatform(imagePlatform, wantedPlatform) {
-				return d.Digest, nil
+		}
+		if (platformCheck && platform.MatchesPlatform(imagePlatform, wantedPlatform)) || (!platformCheck && d.Platform == nil) {
+			if matchedDigest == "" {
+				matchedDigest = d.Digest
+			}
+			if ctx.GivePriorityToZstdImages != types.OptionalBoolFalse {
```

Comment:
maybe not enough tea yet, but should there be an else here?  It looks like we will at best just set `matchedDigest` to `d.Digest` again in this if....

Reply:
@TomSweeneyRedHat This is needed to preserve the original logic of image selection , also added a test to verify that we preserve the original logic. But please unresolve if you think something is missing :)

Reply:
Just to make sure we're talking the same change, my suggestion is:

```
		if matchedDigest == "" {
				matchedDigest = d.Digest
		} else {
			     if ctx.GivePriorityToZstdImages != types.OptionalBoolFalse {
                                // Always give priority to zstd if more than one image
                  	       // exists for the same platform.
				if _, ok := d.Annotations[OCI1InstanceAnnotationCompressionZSTD]; ok {
					matchedDigest = d.Digest
					return matchedDigest
				}
			}
		}
```

(spacing, tabs, brackets way off).  It's not much of a times saving, but I don't think `d.Digest` changes at all between the two lines that assign its value to matchedDigest, so no need to spend time on the two if statements if matchedDigest is already set.

But perhaps I'm misreading?